data/transactions/logic/eval.go

```diff
@@ -934,14 +956,17 @@ func (cx *EvalContext) step() error {
 
 	if err == nil {
 		postheight := len(cx.stack)
-		if postheight-preheight != len(spec.Returns)-len(spec.Args) && spec.Name != "return" {
+		if postheight-preheight != len(spec.Return.Types)-len(spec.Arg.Types) && !spec.AlwaysExits() {
 			return fmt.Errorf("%s changed stack height improperly %d != %d",
-				spec.Name, postheight-preheight, len(spec.Returns)-len(spec.Args))
+				spec.Name, postheight-preheight, len(spec.Return.Types)-len(spec.Arg.Types))
 		}
-		first = postheight - len(spec.Returns)
-		for i, argType := range spec.Returns {
+		first = postheight - len(spec.Return.Types)
+		for i, argType := range spec.Return.Types {
 			stackType := cx.stack[first+i].argType()
 			if !opCompat(argType, stackType) {
+				if spec.AlwaysExits() { // We test in the loop because it's the uncommon case.
```

Comment:
@jannotti Out of curiosity:  Did you observe a benchmarking difference that biased you in the PR's direction?

* The inline comment implies as much.  Asking because it might suggest an intuition I ought to consider while making analogous changes.
* I see that we're in a critical performance code path here.  Though I have little off-hand feel for the cost of branching relative to all other `eval` operations.

Reply:
I did not test this one. I just concluded that calling AlwaysExists and skipping the whole for loop was necessarily slower in the common case - no error - because with no error, you never have to check AlwaysExits at all. By definition, that common case happens on every instruction except, perhaps, the last one.

Reply:
@jannotti Got it - Thanks for the explanation.  Resolving.